### PR TITLE
[DYOD] [WIP] Make sort operator multi-column capable and add SortMicroBenchmarks

### PR DESCRIPTION
**Pull request by jonasnoki**

### CODE REVIEW

**Comment by bakoe** [comment]
Here's a comparison of the TPC-H performance **before** applying our changes (commit d2571500dc039dd0e7a483ec2ea6e43b5b381f9d) vs. **after** applying our changes (commit 1abb77e6380416827d7a8fc3f41685d14dac9d6c):

```diff
 +----------------+--------------------+------+--------------------+------+------------+---------+
 | Benchmark      | prev. iter/s       | runs | new iter/s         | runs | change [%] | p-value |
 +----------------+--------------------+------+--------------------+------+------------+---------+
 | TPC-H 01       | 0.6496357321739197 | 39   | 0.6531713604927063 | 40   | +1%        |  0.5957 |
+| TPC-H 02       | 38.89903259277344  | 2335 | 46.932132720947266 | 2816 | +21%       |  0.0000 |
 | TPC-H 03       | 4.560779094696045  | 274  | 4.715278148651123  | 284  | +3%        |  0.0000 |
 | TPC-H 04       | 4.260347366333008  | 256  | 4.208194732666016  | 253  | -1%        |  0.1239 |
+| TPC-H 05       | 2.40301513671875   | 145  | 2.5666661262512207 | 154  | +7%        |  0.0000 |
+| TPC-H 06       | 83.99534606933594  | 6872 | 118.49485778808594 | 7110 | +41%       |  0.2768 |
 | TPC-H 07       | 4.977008819580078  | 299  | 5.2055439949035645 | 313  | +5%        |  0.0000 |
 | TPC-H 08       | 3.370591640472412  | 203  | 3.4964518547058105 | 210  | +4%        |  0.0000 |
 | TPC-H 09       | 1.1586252450942993 | 70   | 1.1793214082717896 | 71   | +2%        |  0.2046 |
 | TPC-H 10       | 1.8910139799118042 | 114  | 1.847568154335022  | 111  | -2%        |  0.0304 |
 | TPC-H 11       | 18.009695053100586 | 1081 | 18.234140396118164 | 1095 | +1%        |  0.0000 |
 | TPC-H 12       | 9.254011154174805  | 556  | 9.594390869140625  | 576  | +4%        |  0.0000 |
+| TPC-H 13       | 1.0624150037765503 | 64   | 1.4652011394500732 | 88   | +38%       |  0.0000 |
 | TPC-H 14       | 31.235410690307617 | 1875 | 29.74830436706543  | 1785 | -5%        |  0.0000 |
 | TPC-H 15       | 62.0716438293457   | 3725 | 61.88772201538086  | 3714 | -0%        |  0.2110 |
+| TPC-H 16       | 4.684607982635498  | 282  | 5.617599964141846  | 338  | +20%       |  0.0000 |
+| TPC-H 17       | 2.646641492843628  | 159  | 2.825106143951416  | 170  | +7%        |  0.0000 |
 | TPC-H 18       | 0.7338114976882935 | 45   | 0.7413405179977417 | 45   | +1%        |  0.6798 |
 | TPC-H 19       | 4.797001838684082  | 288  | 4.923618793487549  | 296  | +3%        |  0.0010 |
 | TPC-H 20       | 16.51645278930664  | 992  | 16.119413375854492 | 968  | -2%        |  0.0000 |
 | TPC-H 21       | 0.7722448110580444 | 47   | 0.7365230917930603 | 45   | -5%        |  0.0014 |
+| TPC-H 22       | 8.162626266479492  | 490  | 8.752893447875977  | 526  | +7%        |  0.0000 |
+| geometric mean |                    |      |                    |      | +6%        |         |
 +----------------+--------------------+------+--------------------+------+------------+---------+
```

**Comment by mrks** [comment]
This looks like you are including improvements that came from merging the master, too? It should be easier to run `./hyriseBenchmarkTPCH -o foo.json` on both the master and your branch and use `compare_benchmarks.py` to create a comparison.

**Comment by bakoe** [comment]
@mrks Oh yes, that’s right. I’ll do another comparison based on the way you described.

**Comment by mrks** [comment]
The hyriseBenchmarkTPCH binary also uses the sort to prepare the tables. The faster this is, the more convenient it is to run benchmarks.  You might want to add that to your list of contributions,  too.

**Comment by bakoe** [comment]
Updated TPC-H performance comparison of the **master** branch (d2571500dc039dd0e7a483ec2ea6e43b5b381f9d) vs. our current PR state (1abb77e6380416827d7a8fc3f41685d14dac9d6c):

```diff
 +----------------+--------------------+------+--------------------+------+------------+---------+
 | Benchmark      | prev. iter/s       | runs | new iter/s         | runs | change [%] | p-value |
 +----------------+--------------------+------+--------------------+------+------------+---------+
 | TPC-H 01       | 0.6615901589393616 | 40   | 0.6531713604927063 | 40   | -1%        |  0.2005 |
+| TPC-H 02       | 40.73917007446289  | 2445 | 46.932132720947266 | 2816 | +15%       |  0.0000 |
 | TPC-H 03       | 4.546040058135986  | 273  | 4.715278148651123  | 284  | +4%        |  0.0000 |
 | TPC-H 04       | 4.210179328918457  | 253  | 4.208194732666016  | 253  | -0%        |  0.9496 |
+| TPC-H 05       | 2.4139575958251953 | 145  | 2.5666661262512207 | 154  | +6%        |  0.0000 |
 | TPC-H 06       | 115.94351959228516 | 6957 | 118.49485778808594 | 7110 | +2%        |  0.0000 |
 | TPC-H 07       | 5.204839706420898  | 313  | 5.2055439949035645 | 313  | +0%        |  0.9856 |
 | TPC-H 08       | 3.62990665435791   | 218  | 3.4964518547058105 | 210  | -4%        |  0.0000 |
 | TPC-H 09       | 1.1806957721710205 | 71   | 1.1793214082717896 | 71   | -0%        |  0.8850 |
 | TPC-H 10       | 1.8862420320510864 | 114  | 1.847568154335022  | 111  | -2%        |  0.0684 |
 | TPC-H 11       | 18.202131271362305 | 1093 | 18.234140396118164 | 1095 | +0%        |  0.4789 |
 | TPC-H 12       | 9.468599319458008  | 569  | 9.594390869140625  | 576  | +1%        |  0.0000 |
-| TPC-H 13       | 1.5981374979019165 | 96   | 1.4652011394500732 | 88   | -8%        |  0.0000 |
-| TPC-H 14       | 31.862634658813477 | 1912 | 29.74830436706543  | 1785 | -7%        |  0.0000 |
 | TPC-H 15       | 62.11576843261719  | 3727 | 61.88772201538086  | 3714 | -0%        |  0.1134 |
+| TPC-H 16       | 4.856360912322998  | 292  | 5.617599964141846  | 338  | +16%       |  0.0000 |
+| TPC-H 17       | 2.6693525314331055 | 161  | 2.825106143951416  | 170  | +6%        |  0.0000 |
-| TPC-H 18       | 0.7827396392822266 | 47   | 0.7413405179977417 | 45   | -5%        |  0.0205 |
 | TPC-H 19       | 4.921579837799072  | 296  | 4.923618793487549  | 296  | +0%        |  0.9590 |
 | TPC-H 20       | 16.069896697998047 | 965  | 16.119413375854492 | 968  | +0%        |  0.5094 |
 | TPC-H 21       | 0.7259087562561035 | 44   | 0.7365230917930603 | 45   | +1%        |  0.2648 |
 | TPC-H 22       | 8.676556587219238  | 521  | 8.752893447875977  | 526  | +1%        |  0.0503 |
 | geometric mean |                    |      |                    |      | +1%        |         |
 +----------------+--------------------+------+--------------------+------+------------+---------+
```

**Comment by tobodner** [comment]
IMHO, the SortMicroBenchmarks are sufficiently independent to be submitted in a separate PR.

**Comment by tobodner** [comment]
This PR seems to overlap in part with PR #1987.

File: src/lib/operators/sort.hpp [hunk 1]
```diff
@@ -14,19 +14,28 @@
...
+struct SortColumnDefinition final {
+  SortColumnDefinition(const ColumnID& column, const OrderByMode order_by_mode = OrderByMode::Ascending)
+      : column(column), order_by_mode(order_by_mode) {}
+
+  const ColumnID column;
+  const OrderByMode order_by_mode;
+};
+
 /**
  * Operator to sort a table by a single column. This implements a stable sort, i.e., rows that share the same value will
```
Comment (tobodner) [inline src/lib/operators/sort.hpp]: Update this part of the comment as well.

**Comment by tobodner** [comment]
It looks like your CI issues are related to PR #2002.

File: src/lib/operators/sort.hpp (hunk 1 above)
Comment (bakoe) [reply src/lib/operators/sort.hpp]: Resolved in cadf5d739cfc4a713f547ff17b231d1bef9db20f

**Pull request closed without merging**